Social Link: group all variations under one block type

mcsf opened the pull request:
Fixes #17277.

## Background

Initially, Social Links were defined (#16897) as a multitude of block types (`core/social-link-spotify`, `core/social-link-twitter`) all based on a mold of a single social link. The block types would allow only one parent, the container `core/social-links`. From a user standpoint this works, but there are downsides:
- In practice, this meant that the number of block types would grow with Social Links' number of supported sites.
- The document semantics arguably become too specialised, since `<!-- wp:social-link {"site":"wordpress","url":"https://example.org"} /-->` is descriptive and more generalisable.
- Because Social Links are currently defined as server-rendered blocks, there is overhead in registering all these variants on the server before page render. _Making Social Links static is a discussion to be had in #19888._
- From the previous point, it follows that REST API responses for the block type endpoints also increase in size to accommodate the generated block types.

## Description

Recently, the idea of Block Variations (#16283) has solidified (#19243). Initially oriented towards patterns of (inner) blocks, Variations can now accommodate more varied scenarios. One such scenario is that of Social Links.

This PR reimplements Social Links as made up one parent, `core/social-links`, and one child, `core/social-link`. The child then defines a number of variations, one per supported site. This represents no noticeable difference for the user, except for the change in the content markup:

Before|After
-|-
`<!-- wp:social-link-wordpress {"url":"https://example.org"} /-->`|`<!-- wp:social-link {"url":"https://example.org","site":"wordpress"} /-->`

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
![social-links-block-variations](https://user-images.githubusercontent.com/150562/73346797-61106b80-427e-11ea-88d1-053ef60c8a2f.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

gziolo reviewed (commented):
It looks great. Is there anything missing in the API?

I guess you should use store to get patterns to make this block extensible.

mcsf commented on packages/block-library/src/social-link/index.js:
```diff
@@ -7,42 +7,30 @@ import { __ } from '@wordpress/i18n';
 		html: false,
 	},
 	edit,
+	description: __( 'Link to a social media profile' ),
```
I'm not very happy with this description, as the patterns offer many kinds of Web-based resources: social media profiles, multimedia content, misc. pages.

mcsf commented on packages/block-library/src/social-link/index.php:
```diff
@@ -92,6 +92,24 @@ function register_block_core_social_link() {
 			)
 		);
 	}
+
```
I'd love to get rid of the loop above right away, in which all `core/social-link-<foo>` types are generated, but in order to do that we need a better plan for back-compat.

Right now only the editor has a back-compat provision via the change in `packages/blocks/src/api/parser.js`. How could we best handle this for server-side block rendering for the front-end?

gziolo replied:
You should check the HTML output of the edit post page. All blocks registered here will have a huge impact on the size of the page. In particular, the part that registers blocks to work with the `ServerSideRender` component (which is also pointless in this case as they will never be consumed).

At the same time, it's something that was never part of the WordPress core so we can provide the backward-compatibility only for the Gutenberg plugin and call it the day 😄

mcsf replied:
> At the same time, it's something that was never part of the WordPress core so we can provide the backward-compatibility only for the Gutenberg plugin and call it the day 😄

Yeah, that seems like the right way to go. Let's address in a subsequent PR.

mcsf replied:
#20098

mcsf commented on packages/block-library/src/social-link/patterns.js:
```diff
@@ -0,0 +1,285 @@
+		attributes: { site: 'wordpress' },
+		title: 'WordPress',
+		icon: WordPressIcon,
+	},
```
Unlike the Embed block, the Social Link block was initially conceived so that it can't be "generic", i.e. it has to follow a pattern. Given its purpose, this makes sense.

On the other hand, the way that Block Variations / Patterns are currently defined, if the "base" type is to be supplanted by the patterns there needs to be a default one. I chose _WordPress_ as the least opinionated one (or the least controversial opinion?).

I guess we could instead adapt the Patterns API so that it's possible to supplant the base type without requiring a default, but I'm not sure it's worth adding that compromise. What do you think, @gziolo and @mtias?

mcsf replied:
Side note: gotta love Vim for how magically it was able to transform `social-list.js` into `patterns.js` with just a quick macro. <3

mcsf commented:
> I guess you should use store to get patterns to make this block extensible.

That's a good point. It's also something we may want to improve: shouldn't statically defined patterns be extensible or removable at runtime?

mcsf commented:
> It looks great. Is there anything missing in the API?

Right now the only thing that comes to mind is https://github.com/WordPress/gutenberg/pull/19887#discussion_r371296884

mkaz commented:
I'm wondering if Social Links should be just an Icon package for the logos, then the actual implementation of a social menu would look more like the Buttons block or Navigation block. This would do away with generating all of the individual site blocks.

Along the same lines, we could not have a special social links block and just add icon support to Buttons and/or Navigation.

paaljoachim commented:
Marcus that sounds very interesting!
Having the social blocks use the Buttons Block implementation seems like a good idea.
In a sense there could be two almost identical blocks. Buttons and Social. The user might find it easier having a Buttons Block and a Social Block, even though they are almost identical.
@jasmussen

jasmussen commented:
> Along the same lines, we could not have a special social links block and just add icon support to Buttons and/or Navigation.

Purely on a technical and code quality level, you all should do what you need to do! If navigation, social links and buttons are technically a single block, cool with me. It even makes sense insofar as one might very well want both navigation items AND a couple of social link icons in the very same navigation block.

However from a user and block library discoverability perspective, though, it is important that Navigation, Social Links and Buttons remain three separate blocks, though, so that you can search the block library for either and insert it with confidence.

Like I said, it could be like the embed block where the YouTube block is literally the same block as the Embed block, just with a different name and icon. But it's important that they each have a separate entry in the block library. Why? This goes all the way back to [the start](https://make.wordpress.org/core/2017/01/04/focus-tech-and-design-leads/):

> The editor will endeavour to create a new page and post building experience that makes writing rich posts effortless, and has "blocks" to make it easy what today might take shortcodes, custom HTML, **or "mystery meat" embed discovery**.

Emphasis mine.

gziolo commented:
> However from a user and block library discoverability perspective, though, it is important that Navigation, Social Links and Buttons remain three separate blocks, though, so that you can search the block library for either and insert it with confidence.

This is exactly what we discussed on WordPress Slack with @mkaz and @mcsf yesterday (link requires registration):
https://wordpress.slack.com/archives/C02QB2JS7/p1580308195116200

We want to make it possible with new block variations (formerly patterns) API.

mkaz commented:
For backward compatibility, can we add a deprecation to the social-links parent and use a migrate function which receives both attributes and InnerBlocks, and convert the InnerBlocks to the new social-link type. I think this can solve it in editor.

I'm not sure how we'd deprecate the server-side rendering

mcsf committed:
Social Link: group all variations under one block type

mcsf commented:
> For backward compatibility, can we add a deprecation to the social-links parent and use a migrate function which receives both attributes and InnerBlocks, and convert the InnerBlocks to the new social-link type. I think this can solve it in editor.

This was solved early on by tackling the issue [right in the parser](https://github.com/WordPress/gutenberg/pull/19887/files#diff-7dae7b3890cf27659680c926b6992e31), following an established patterns of renamed blocks (e.g. `core/text`).

mcsf committed:
Social Link: Update test fixtures

gziolo commented on packages/block-library/src/social-link/edit.js:
```diff
@@ -33,8 +33,8 @@ const SocialLinkEdit = ( { attributes, setAttributes, isSelected } ) => {
-	const IconComponent = getIconBySite( site );
-	const socialLinkName = getNameBySite( site );
+	const IconComponent = site ? getIconBySite( site ) : null;
+	const socialLinkName = site ? getNameBySite( site ) : '';
```
Should it default to: `_( 'Social Icon' )` or use directly the name of the block?

```suggestion
	const socialLinkName = site ? getNameBySite( site ) : _( 'Social Icon' );
```

mcsf replied:
That makes sense, or `Social Link` for consistency?

mcsf replied:
The title of the block itself is _Social Icon_, so I'll use that for now.

gziolo commented on packages/block-library/src/social-link/index.js:
```diff
@@ -7,40 +7,30 @@ import { __ } from '@wordpress/i18n';
-				},
-			},
+	description: __( 'Create a link to the wider Web' ),
+	attributes: {
```
Nit: now that it is only a single block, `block.json` can be used.

gziolo commented on packages/block-library/src/social-link/index.php:
```diff
@@ -92,6 +92,24 @@ function register_block_core_social_link() {
+
+	register_block_type(
+		'core/social-link',
+		array(
```
With `block.json` in place, the following is possible:

https://github.com/WordPress/gutenberg/blob/2c587e5e252852870c087c1d9874d8867aa53c6d/packages/block-library/src/shortcode/index.php#L24-L34

It helps to avoid syncing attributes between PHP and JS. WordPress core has everything in place to handle it.

mcsf replied:
True, thanks!

gziolo commented on packages/block-library/src/social-link/social-list.js:
```diff
@@ -1,227 +1,31 @@
-export const getIconBySite = ( site ) => {
-	return socialList[ site ].icon;
+export const getIconBySite = ( name ) => {
+	return find( variations, { name } ).icon;
```
It needs to be guarded in the case where a custom variation is registered and the site can't be found.

gziolo commented on packages/block-library/src/social-link/variations.js:
```diff
@@ -0,0 +1,285 @@
+	YouTubeIcon,
+} from './icons';
+
+const patterns = [
```
```suggestion
const variations = [
```
:)

mcsf replied:
😬

gziolo commented on packages/block-library/src/social-links/edit.js:
```diff
@@ -4,23 +4,16 @@
-const ALLOWED_BLOCKS = Object.keys( socialList ).map( ( site ) => {
-	return 'core/social-link-' + site;
-} );
+const ALLOWED_BLOCKS = [ 'core/social-link' ];
```
Nice :)

gziolo commented on packages/blocks/src/api/parser.js:
```diff
@@ -429,6 +429,14 @@ export function createBlockWithFallback( blockNode ) {
 		name = 'core/paragraph';
 	}
 
+	// Convert derivative blocks such as 'core/social-link-wordpress' to the
```
Yes, this should be enough for now. In the long term, we should consider a filter that allows registering those handlers.  

```js
// Convert derivative blocks such as 'core/social-link-wordpress' to the
// canonical form 'core/social-link'.
const convertLegacySocialLink = ( { name, attributes } ) => {
    if ( name && name.indexOf( 'core/social-link-' ) === 0 ) {
        // Capture `social-link-wordpress` into `{"site":"wordpress"}`
        attributes.site = name.substring( 17 );
        name = 'core/social-link';
    }
    return { name, attributes };
}
```

You would register this type of filter just before the block gets registered.

gziolo reviewed (commented):
I had some minor notes, but overall I like how it simplifies things. Code changes look good. It needs some testing, polishing and we should include it in WordPress 5.4 👍 

I also shared the initial proposal of how we could handle block name deprecations. It's something that should be explored separately though.

mcsf committed:
Address review feedback

mcsf commented:
Thanks for the great feedback, I've addressed everything.

mtias commented:
I 100% agree with this direction. I don't think an abstraction such as "icon" or "buttons" would be a great idea, as it moves too far and loses semantically. One thing to consider is if `social-link` is better than `social-icon` for the purpose of this block.

mtias commented on packages/block-library/src/social-link/index.js:
```diff
@@ -7,40 +7,21 @@ import { __ } from '@wordpress/i18n';
 		html: false,
 	},
 	edit,
+	description: __( 'Create a link to the wider Web' ),
```
This description could be improved. Reading it in isolation doesn't hint what the block does.

mcsf replied:
Yes, same. My initial diff had a comment about it. Ideas?

mkaz replied:
Create an icon link to external sites.

mcsf replied:
I'll try _Display an icon linking to a social media profile or website_

mtias commented on packages/block-library/src/social-links/edit.js:
```diff
@@ -4,23 +4,16 @@
-	[ 'core/social-link-instagram' ],
-	[ 'core/social-link-linkedin' ],
-	[ 'core/social-link-youtube' ],
+	[ 'core/social-link', { site: 'wordpress', url: 'https://wordpress.org' } ],
```
This looks a lot cleaner.

mtias commented on packages/block-library/src/social-link/block.json:
```diff
@@ -0,0 +1,15 @@
+		"url": {
+			"type": "string"
+		},
+		"site": {
```
Is "site" the best attribute name?

mtias replied:
What about `service`?

mcsf replied:
I've wondered about `service` too. Could try that.

mcsf replied:
How's this, @mtias? ba9070f

mtias commented on packages/block-library/src/social-links/index.js:
```diff
@@ -25,16 +25,25 @@ export const settings = {
-				attributes: { url: 'https://wordpress.org' },
+				name: 'core/social-link',
+				attributes: {
+					type: 'wordpress',
```
Is this supposed to be "type" or "site"?

mcsf replied:
Good catch.

mcsf committed:
Fix misnamed attribute 'type' -> 'site'

mcsf committed:
Social Link: Rename attribute 'site' to 'service'

mcsf committed:
Social Link: Update description

mkaz commented:
It looks like you're still updating the PR, but changing the attribute to `service` requires updates in edit.js and index.php for getting the icons to render properly. I don't mind the change, but concerned we're introducing another term, the code is already a mix of site, icon, and link.

mkaz committed:
Change attribute name to service (previously site)

mkaz commented:
I pushed up an update to fix attribute change to service. A code quality improvement would probably be to change the function names to use service also. This can wait because I want to refactor how the SVG are displayed in #19888 and I might be changing functions.

mkaz reviewed (approved):
Great work :+1:

mkaz committed:
Lint: Equal sign alignment

mcsf commented:
> I pushed up an update to fix attribute change to service.

Thanks! Yes, this was a "commit-and-groceries" kind of dev flow. 😬 

> A code quality improvement would probably be to change the function names to use service also. This can wait because I want to refactor how the SVG are displayed in #19888 and I might be changing functions.

Agreed.

Thanks for the review!

mcsf committed:
Social Link: Remove unnecessary block deprecation

This PR introduces `core/social-link` as a new block type. Strictly
speaking, any concerns of backwards compatibility should only deal with
the migration of `core/social-link-FOO` types to this new one --
something that the parser is handling directly, populating the new
`service` attribute in that process.

Thus, no specific provisions are needed for the extant but obsolete
`site` attribute: the attribute will be silently dropped during block
validation. This can be attested by loading any content with links in
the old `core/social-link-FOO` shape.

mcsf commented:
Before I merge, I pushed 07a9e36. Quoting the commit message:

> Remove unnecessary block deprecation
> 
> This PR introduces `core/social-link` as a new block type. Strictly speaking, any concerns of backwards compatibility should only deal with the migration of `core/social-link-FOO` types to this new one — something that the parser is handling directly, populating the new `service` attribute in that process.
> 
> Thus, no specific provisions are needed for the extant but obsolete `site` attribute: the attribute will be silently dropped during block validation. This can be attested by loading any content with links in the old `core/social-link-FOO` shape.

I'll wait for the tests to pass.

gziolo reviewed (commented):
I haven't tested myself all this but code-wise it should be good to go once https://github.com/WordPress/gutenberg/pull/19887/files#r375784004 is addressed. It was one of the primary reasons why this block didn't land earlier in the WordPress core.

gziolo commented on packages/block-library/src/social-link/block.json:
```diff
@@ -0,0 +1,15 @@
+{
+	"name": "core/social-link",
```
I'm also wondering now that I noticed that the block's title is now `Social Icon`, should we change the name as well? :)

mcsf replied:
When I started this PR I was pretty much on the `links` camp, but now I lean a bit towards `icon` too.

ZebulanStanphill replied:
@mcsf Why?

mcsf replied:
Because I was initially considering the primary function of these blocks to be the linking — and this fit in with the possibility of declaring Social Links as a variation of Navigation or Buttons. But now I think we may compromise on functionality if we center the semantics of these blocks around just the linking.

Instead, their real purpose lies more in the appearance: the rendering of the links as icons. In the future, the icons may change looks, colours or shape, but they will remain icons. I think the user is best served if we are clear about this. If, instead, the primary purpose is linking, then other blocks are better suited (Navigation, Buttons, File, etc.).

mcsf replied:
I have a diff that moves everything to Icons, but it's very large, so I'll open a separate PR to discuss this.

gziolo replied:
Sure thing 💯

ZebulanStanphill replied:
The point of having social icon links is that they're links, not that they're icons, isn't it? From an accessibility standpoint, I question whether we should even have a block dedicated to textless buttons in the first place. Wouldn't that be sort of an anti-pattern?

Also, what if I want to have social links with visible text labels? Would I use the Buttons block for that?

mcsf replied:
> The point of having social icon links is that they're links, not that they're icons, isn't it?

That's the subject part of the debate, I think.

> From an accessibility standpoint, I question whether we should even have a block dedicated to textless buttons in the first place. Wouldn't that be sort of an anti-pattern?

If done correctly, none of it should matter: icons should be as accessible as a regular text link, whether the reader cannot see, cannot use a cursor, etc.

> Also, what if I want to have social links with visible text labels? Would I use the Buttons block for that?

I could see either block type accommodating that use case.

I'll cc you on the new PR if you're interested in discussing this there.

gziolo reviewed (approved):
I processed my previous comments and this is my conclusion, Let's land and improve block registration in PHP later. I don't see it as a blocker for the beta. In the plugin, it's going to work the same way anyway so we have a month to improve it :)
